### ISSUE TITLE
No nullable warning when combining static generic fields and nullable generic type instantiations

### ISSUE THREAD

**Issue by T-Gro** [dotnet/roslyn#75677]
**Version Used**:  Main from 11 Jun 2024

**Steps to Reproduce**:

The basic idea comes from the design of nullable reference types being erased and  unknown to the runtime.
With a static storage in a generic type, the data for `C<string>` and `C<string|null>` is shared. However, this is neither understood nor warned about.

```
using System;
using System.Collections.Generic;

C<String>.Instance[0] = "hi";
C<String?>.Instance[0] = null;
Console.Write(C<String>.Instance[0].Length);

public class C<T> {
    public static Dictionary<int,T> Instance {get;set;}
    static C() {
        Instance = new Dictionary<int, T>();
    }
}
```

https://sharplab.io/#v2:D4AQTAjAsAUCAMACEEAsBuWsDCAeAygC4BOAlgHYDmAfAHQCS5AzoQIbkDGApgNrwC6iALyIARAAtSozDDxEyVAPx1GLdtz6CR5AK4AbPTOwB7Zsb1daAdTKEuACjkkKNBszadeA2gBkuVQnEAShlYEABmZDBEPAAVakQAb1hEVORIlAA2RAARUg5CUlNWYgBPXApCABp4xFUPbiTKLkJ0Jhb0AF8UtKyY+yCknrS0+vUuYURyLgB3XPzC4rKK8mrEeIGZEe6YHawYIA

**Expected Behavior**:
**Any** kind of warning on using static data in generics with a nullable type instantiation.

**Actual Behavior**:
System.NullReferenceException: Object reference not set to an instance of an object.

**Comment by jjonescz** [comment]
I don't think nullability analysis can do much here (it would either need to know whether there are some `C<string?>` usages in any method body anywhere or it would warn people having only `C<string>` unnecessarily). But I think you can decide yourself if `C<T>` should be usable with nullable reference types or not and then either
- add `class C<T> where T : notnull` to warn on the `C<string?>` usage; or
- declare as `Dictionary<int,T?>` which will cause warning on `C<String>.Instance[0].Length` access.

**Comment by T-Gro** [comment]
This is specifically about generic types explicitly allowing **both** nullable and not nullable type instances.
(also considering the case where owner of the generic type and its consumers are different people/teams/libraries..).

It should be kept in mind that such usage can be fairly accidental.

What about a new nullable warning telling about type erasure when a nullable type is used to access static data in generic type? 
"Warning: The type 'string?' is used to access static generic data of a class. This will collide with usage using 'string' as the type argument. Learn more at.."

**Comment by RikkiGibson** [comment]
I think if we did anything here, the only reasonable place to warn would be on the field/auto-prop declaration itself.

It's not clear to me if such a warning would be excessively disruptive. So, I can't say whether it's realistic for us to take any change to address this scenario.

I would be open to see somebody propose a precise set of conditions to warn on, attempt to add the warning, and seeing what kind of impact it creates around the ecosystem (e.g. in building repos like roslyn, runtime, major third party libraries and so on.)

**Comment by T-Gro** [comment]
> precise set of conditions to warn on


This is the part where it get's tricky, because the definition and the two conflicting usages of the generic type can be in 3 different compilations.
(let's say:
LibraryA defines the generic type
LibraryB uses the generic type with "string" type instance
App references LibraryA and LibraryB , and uses the type with "string|null" type instance.)


I think Roslyn understands this concept when within the same method, but doing this across whole compilations (or even more of them) is another challenge.


This is the scenario where Roslyn can detect that the two statics are actually the same:
https://sharplab.io/#v2:D4AQTAjAsAUCAMACEEAsBuWsDCAeAygC4BOAlgHYDmAfAHQCS5AzoQIbkDGApogLyIAiABakBmGNgD2zSQBsutAOplCXABR4iZKnUYt23WgBkuVQkICU6RAHobiAHKTEAd1bFyFSlhg4CJLwB+XWY2Th5+cgBXWVlxKRl5JRV1TQCdBlCDBRMzS2tbewAlSSZZAE9yRABrckkXJkRzVkImySiOESpENO1KYMRWADMhrg5VABMe/z7qHxAAZmQwaYAVakQAb1hEXeQllAA2RFXEPTDuLcouQnQmG/QAXx29l92jnrULN62fvbOsuE+IgJlwhqwYoQAITif7PXwweE+IA=

**Comment by RikkiGibson** [comment]
Yes, within a single method, nullable analysis simply tracks a flow state for the static field. It probably will end up using the generic type argument to determine an initial state for it per-method.

We try hard to avoid any interprocedural analysis in the compiler in general. But even moreso when it is compilation-wide or even cross-compilation. That is the domain of third-party tools.

Hence my suggestion that if we want to give a warning here, it needs to be reported on the field declaration, and based solely on the field declaration.
